### PR TITLE
[FIRRTL] Update LowerClasses to handle extmodule and instances.

### PR DESCRIPTION
**Pull request by mikeurbach**
This adds lowering for extmodule similar to extclasses: an external class declaration is created with the property input and output ports. This also adds support for converting instances of modules or extmodules that have been lowered, and keeping the instance graph up to date.

A small addition is also included to always create a class for public modules and extmodules, which is a further refinement of the as-yet undocumented ABI.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 1]
```diff
@@ -247,15 +253,19 @@ void LowerClassesPass::runOnOperation() {
...
 
   // Convert to OM ops and types in Classes or Modules in parallel.
   if (failed(
           mlir::failableParallelForEach(ctx, objectContainers, [&](auto *op) {
             return dialectConversion(op, pathInfoTable);
           })))
     return signalPassFailure();
+
+  // Mark analyses preserved, since we keep the instance graph up to date.
+  markAllAnalysesPreserved();
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:268]: This seems like too aggressive. Please preserve only instance graph analysis.

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 2]
```diff
@@ -440,92 +457,239 @@ void LowerClassesPass::lowerClassExtern(ClassExternOp classExternOp,
...
-    unsigned nextArgIndex = 0;
-    for (unsigned i = 0; i < numElements; ++i) {
-      auto direction = firrtlClassType.getElement(i).direction;
-      if (direction == Direction::In)
-        argIndexTable[i] = nextArgIndex++;
-    }
+// Helper to update an Object instantiation. FIRRTL Object instances are
+// converted to OM Object instances.
+static LogicalResult
+updateObjectInstance(firrtl::ObjectOp firrtlObject, OpBuilder &builder,
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:463]: This should be the same implementation as what was previously in the loop body of `updateInstances`. While resolving merge conflicts, I just copy-pasted the implementation currently on main. The only change I made was to delay erasing ops, so the main loop in updateInstances wouldn't visit an up that had already been deleted.

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 3]
```diff
@@ -770,11 +934,11 @@ struct ObjectOpConversion : public OpConversionPattern<om::ObjectOp> {
...
   matchAndRewrite(om::ObjectOp objectOp, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
-    rewriter.updateRootInPlace(objectOp, [&] {
-      auto operands = adaptor.getOperands();
-      for (unsigned i = 0, e = objectOp->getNumOperands(); i < e; ++i)
-        objectOp->setOperand(i, operands[i]);
-    });
+    // Replace the object with a new object using the converted actual parameter
+    // types from the adaptor.
+    rewriter.replaceOpWithNewOp<om::ObjectOp>(objectOp, objectOp.getType(),
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:939]: While resolving merge conflicts, I took the implementation I had over the one on main. I think it does the same thing, but the one-liner is a little simpler. This change isn't technically required, and I am also happy to back it out to keep this patch simpler.

**Review by mikeurbach — COMMENTED** [review]
Since there were some merge conflicts, could you take a quick look @rwy7?

**Review by rwy7 — APPROVED** [review]

**Pull request merged**